src/NUnitTestAdapter/NUnit3TestExecutor.cs

```diff
@@ -209,14 +207,18 @@ private void RunAssembly(string assemblyName, TestFilter filter)
                 Debugger.Launch();
 #endif
 
+            var actionText = Debugger.IsAttached ? "Debugging " : "Running ";
+            var selectionText = filter == null || filter == TestFilter.Empty ? "all" : "selected";
+            TestLog.Info(actionText + selectionText + " tests in " + assemblyName);
+
             // No need to restore if the seed was in runsettings file
             if (!Settings.RandomSeedSpecified)
                 Settings.RestoreRandomSeed(Path.GetDirectoryName(assemblyName));
 
-            _activeRunner = GetRunnerFor(assemblyName);
-
             try
             {
+                _activeRunner = GetRunnerFor(assemblyName);
```

Comment:
This and your finally clause below appear to be the fix for the issue, but as @CharliePoole says, you have included numerous unrelated changes. This code looks fine and the unrelated `TestLog` changes are probably fine, but they are unrelated, so it makes it harder to accept this PR.